atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/atomic/KnowledgeableTimestampExtractingAtomicTable.java

```diff
@@ -0,0 +1,116 @@
+    }
+
+    @Override
+    public void markInProgress(Iterable<Long> keys) {
+        delegate.markInProgress(keys);
+    }
+
+    @Override
+    public void updateMultiple(Map<Long, Long> keyValues) throws KeyAlreadyExistsException {
+        delegate.updateMultiple(KeyedStream.stream(keyValues)
+                .map(TransactionStatusUtils::fromTimestamp)
+                .collectToMap());
+    }
+
+    /**
+     * Returns commit timestamp for the start timestamp supplied as arg.
+     * For transactions that are successfully committed, returns the respective commit timestamps.
+     * For transactions that are aborted, returns -1.
+     * For transactions that are unknown, returns startTs as commitTs for read-write transactions.
+     * For read-only transactions, only returns if the greatestSeenCommitTS < startTs, otherwise throws.
+     * Start timestamps for transactions that are in progress return a void future.
+     * */
+    @Override
+    public ListenableFuture<Long> get(Long startTimestamp) {
+        return getInternal(startTimestamp);
+    }
+
+    /**
+     * Returns commit timestamps against start timestamps supplied as arg.
+     * Delegates to {@link KnowledgeableTimestampExtractingAtomicTable#get(Long)} serially.
+     * Start timestamps for transactions that are in progress are not included in the map.
+     * */
+    @Override
+    public ListenableFuture<Map<Long, Long>> get(Iterable<Long> keys) {
+        Map<Long, ListenableFuture<Long>> futures = KeyedStream.of(
+                        StreamSupport.stream(keys.spliterator(), false).sorted(Comparator.reverseOrder()))
+                .map(this::getInternal)
+                .collectToMap();
+        return AtlasFutures.allAsMap(futures, MoreExecutors.directExecutor());
+    }
+
+    @VisibleForTesting
+    ListenableFuture<Long> getInternal(long startTimestamp) {
+        if (knownConcludedTransactions.isKnownConcluded(
+                startTimestamp, KnownConcludedTransactions.Consistency.LOCAL_READ)) {
+            return Futures.immediateFuture(getCommitTsForConcludedTransaction(startTimestamp));
+        } else {
+            ListenableFuture<TransactionStatus> presentValueFuture = delegate.get(startTimestamp);
+            return Futures.transform(
+                    presentValueFuture,
+                    presentValue -> getCommitTsFromStatus(startTimestamp, presentValue),
+                    MoreExecutors.directExecutor());
+        }
+    }
+
+    private Long getCommitTsFromStatus(long startTs, TransactionStatus status) {
+        if (status.equals(TransactionStatuses.unknown())) {
+            // unknown status implies that the transactions table has been swept and the transaction is therefore
+            // concluded.
+            return getCommitTsForConcludedTransaction(startTs);
+        } else {
+            return TransactionStatusUtils.maybeGetCommitTs(status).orElse(null);
+        }
```

Comment:
```suggestion
TransactionStatuses.caseOf(status)
                .unknown(() -> getCommitTsForConcludedTransaction(startTs))
                .otherwise(() -> TransactionStatusUtils.maybeGetCommitTs(status).orElse(null));
```